File: src/storage.js
/**
 * In-memory storage repository for device registrations and heartbeats.
 */
export class DeviceStorage {
  constructor(maxHistoryPerDevice = 50) {
    this._devices = new Map();
    this._maxHistory = maxHistoryPerDevice;
  }

  /**
   * Registers a new device.
   * @param {string} id
   * @param {string} name
   * @param {Date} [registeredAt]
   * @returns {object|null} The registered device record, or null if ID already exists.
   */
  registerDevice(id, name, registeredAt = new Date()) {
    if (this._devices.has(id)) {
      return null;
    }

    const record = {
      id,
      name,
      registered_at: registeredAt.toISOString(),
      last_heartbeat: null,
      heartbeat_count: 0,
      latest_telemetry: null,
      heartbeat_history: [],
    };

    this._devices.set(id, record);
    return { ...record };
  }

  /**
   * Retrieves a device record by ID.
   * @param {string} id
   * @returns {object|null}
   */
  getDevice(id) {
    const record = this._devices.get(id);
    if (!record) return null;
    return { ...record };
  }

  /**
   * Returns a snapshot array of all device records.
   * @returns {object[]}
   */
  listDevices() {
    return Array.from(this._devices.values()).map((record) => ({ ...record }));
  }

  /**
   * Records a heartbeat for a registered device.
   * @param {string} id
   * @param {Date} receivedAt
   * @param {object} [telemetry]
   * @returns {object|null} Updated device record, or null if device not found.
   */
  recordHeartbeat(id, receivedAt, telemetry = null) {
    const record = this._devices.get(id);
    if (!record) return null;

    record.last_heartbeat = receivedAt.toISOString();
    record.heartbeat_count += 1;

    if (telemetry && Object.keys(telemetry).length > 0) {
      record.latest_telemetry = { ...telemetry };
      record.heartbeat_history.push({
        timestamp: record.last_heartbeat,
        data: telemetry,
      });

      if (record.heartbeat_history.length > this._maxHistory) {
        record.heartbeat_history.shift();
      }
    }

    return { ...record };
  }

  /**
   * Deletes a device from storage.
   * @param {string} id
   * @returns {boolean} True if removed, false if not found.
   */
  deleteDevice(id) {
    return this._devices.delete(id);
  }

  /**
   * Clears all stored records (useful for test isolation).
   */
  clear() {
    this._devices.clear();
  }

  /**
   * Total number of devices.
   */
  count() {
    return this._devices.size;
  }
}

// Global singleton instance
export const deviceStorage = new DeviceStorage();

File: src/storage.py
from datetime import datetime, timezone
import threading
from typing import Any, Dict, List, Optional


class DeviceRecord:
    """Internal representation of a device in memory."""

    def __init__(self, id: str, name: str, registered_at: Optional[datetime] = None):
        self.id: str = id
        self.name: str = name
        self.registered_at: datetime = registered_at or datetime.now(timezone.utc)
        self.last_heartbeat: Optional[datetime] = None
        self.heartbeat_count: int = 0
        self.latest_telemetry: Optional[Dict[str, Any]] = None
        self.heartbeat_history: List[Dict[str, Any]] = []

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "registered_at": self.registered_at,
            "last_heartbeat": self.last_heartbeat,
            "heartbeat_count": self.heartbeat_count,
            "latest_telemetry": self.latest_telemetry,
        }


class DeviceStorage:
    """Thread-safe in-memory store for device registrations and heartbeats."""

    def __init__(self, max_history_per_device: int = 50):
        self._devices: Dict[str, DeviceRecord] = {}
        self._lock = threading.RLock()
        self._max_history = max_history_per_device

    def register_device(self, id: str, name: str, registered_at: Optional[datetime] = None) -> Optional[DeviceRecord]:
        """Registers a device. Returns the record, or None if device ID already exists."""
        with self._lock:
            if id in self._devices:
                return None
            record = DeviceRecord(id=id, name=name, registered_at=registered_at)
            self._devices[id] = record
            return record

    def get_device(self, id: str) -> Optional[DeviceRecord]:
        """Retrieves a device record by id."""
        with self._lock:
            return self._devices.get(id)

    def list_devices(self) -> List[DeviceRecord]:
        """Returns a snapshot list of all registered devices."""
        with self._lock:
            return list(self._devices.values())

    def record_heartbeat(
        self,
        device_id: str,
        received_at: datetime,
        telemetry: Optional[Dict[str, Any]] = None,
    ) -> Optional[DeviceRecord]:
        """
        Records a heartbeat for a registered device.
        Returns the updated DeviceRecord, or None if the device does not exist.
        """
        with self._lock:
            device = self._devices.get(device_id)
            if not device:
                return None

            device.last_heartbeat = received_at
            device.heartbeat_count += 1
            if telemetry:
                device.latest_telemetry = telemetry
                device.heartbeat_history.append({
                    "timestamp": received_at,
                    "data": telemetry
                })
                # Cap history size
                if len(device.heartbeat_history) > self._max_history:
                    device.heartbeat_history.pop(0)

            return device

    def clear(self) -> None:
        """Clears all devices from memory (primarily for testing)."""
        with self._lock:
            self._devices.clear()

    def count(self) -> int:
        """Returns total registered devices count."""
        with self._lock:
            return len(self._devices)


# Global singleton store
device_storage = DeviceStorage()
